dascore/proc/whiten.py

```diff
@@ -120,127 +247,28 @@ def plot_spectrum(x, T, ax, phase=False):
-        raise ParameterError(msg)
-
-    assert np.all(np.isreal(patch.data)), "Input data needs to be real"
-
-    fft_d = nft.rfft(patch.data, n=comp_nsamp, axis=dim_ind)
-    amp = np.abs(fft_d)  # Original spectrum
-    phase = np.angle(fft_d)  # Original phase
-
-    conv_window = np.arange(mean_window) / mean_window
-
-    # Define tapering window
-    taper_win = np.zeros(nf)
-    taper_win[first_freq_ind:last_freq_ind] = tukey(freq_win_size, alpha=tukey_alpha)
-
-    # convolve original spectrum with smoothing window
-    sm_amp = convolve1d(amp, conv_window, axis=dim_ind, mode="constant")
-
-    # divide original spectrum by smoothed spectrum
-    norm_amp = np.divide(
-        amp, np.abs(sm_amp), out=np.zeros_like(sm_amp), where=abs(sm_amp) != 0
-    )
-
-    # Generate N-D tapering window and multiply normalize spectrum to get desired
-    # frequency range.
-    exp_dims = np.asarray(norm_amp.shape)
-    exp_dims[dim_ind] = 1
-    tiled_win = taper_win
-
-    for d in exp_dims:
-        if d > 1:
-            tiled_win = np.tile(np.expand_dims(tiled_win, axis=-1), (1, d))
-
-    tiled_win = np.transpose(tiled_win, np.roll(np.arange(len(exp_dims)), dim_ind))
-
-    norm_amp *= tiled_win
-
-    if idft:
-        # revert back to time-domain, using the phase of the original signal
-        whitened_data = np.real(
-            nft.irfft(norm_amp * np.exp(1j * phase), n=comp_nsamp, axis=dim_ind)
-        )
-        whitened_data = np.take(whitened_data, np.arange(nsamp), axis=dim_ind)
-        return patch.new(data=whitened_data)
-    else:
-        dims = list(patch.dims)
-        dxs, axes = _get_dx_or_spacing_and_axes(patch, dims, require_evenly_spaced=True)
-        # get new coordinates
-        real = dims[dim_ind]
-        new_coords = _get_dft_new_coords(patch, dxs, dims, axes, real)
-        # get attributes
-        attrs = _get_dft_attrs(patch, dims, new_coords)
-        # return the frequency domain data directly without taking the inverse FFT
-        return patch.new(norm_amp * np.exp(1j * phase), coords=new_coords, attrs=attrs)
+    # Set default kwargs if not set, check inputs.
+    dim, freq_range = _get_dim_freq_range_from_kwargs(patch, kwargs)
+    smooth_size = _check_whiten_inputs(patch, smooth_size, tukey_alpha, dim, freq_range)
+    # Get patch in dft form.
+    fft_patch = patch.dft(dim, real=np.isrealobj(patch.data))
+    input_patch_fft = fft_patch is patch  # if input patch had fft
+    fft_dim = FourierTransformatter().rename_dims(dim)[0]
+
+    # Get an envelope which can be used to normalize the spectra so each freq
+    # has roughly the same amplitude (but phase remains unchanged).
+    envelope = _get_amplitude_envelope(fft_patch, fft_dim, smooth_size, freq_range)
+
+    # Apply the tukey window to the original envelope so only the freqs
+    # in the selected range are used.
+    if freq_range is not None:
+        envelope = _filter_array(envelope, fft_patch, fft_dim, freq_range, tukey_alpha)
```

Comment:
Could you please clarify why we calculate `envelope` first in line 260 and then change it in 265 if the selected range are used?

Reply:
Hey @ahmadtourei, I updated the comment so it is hopefully explained better now.